### PR TITLE
Resizable terminal

### PR DESCRIPTION
**Pull request by larskarlitski**
Add a resize control message to pty channels and make `system/terminal` fullscreen.

### CODE REVIEW

**Comment by stefwalter** [comment]
@andreasn Have you been able to take a look at the layout and design here?

**Comment by stefwalter** [comment]
The integration test is failing. 

```
# testBasic (check_terminal.TestTerminal)
#
Restarting browser...
Wrote TestTerminal-testBasic-FAIL.png
Journal extracted to TestTerminal-testBasic-10.111.125.156-FAIL.log
not ok 66 testBasic (check_terminal.TestTerminal)
Traceback (most recent call last):
  File "./verify/check-terminal", line 58, in testBasic
    self.assertIn(":~$", prompt)
AssertionError: ':~$' not found in u'\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0'
```

File: pkg/base1/cockpit.css [hunk 1]
```diff
@@ -314,15 +314,27 @@ a {
...
+    background-color: black;
+    margin: 0;
     font-size: 10px;
-    text-align: center;
     line-height: 1;
 }
 
+.console-ct-fullscreen {
+    position: absolute;
+    top: 44px;
```
Comment (stefwalter) [inline pkg/base1/cockpit.css]: What is this relative to?

File: pkg/base1/term.css [hunk 2]
```diff
@@ -1,17 +1,3 @@
-.terminal {
-  float: left;
```
Comment (stefwalter) [inline pkg/base1/term.css]: What replaces these styles now?

File: src/bridge/cockpitpipechannel.c [hunk 3]
```diff
@@ -105,6 +107,27 @@ process_pipe_buffer (CockpitPipeChannel *self,
...
 static gboolean
+cockpit_pipe_channel_read_window_size_options (JsonObject *options,
+                                               gushort default_rows,
+                                               gushort default_cols,
+                                               gushort *rowsp,
+                                               gushort *colsp)
+{
+  gint64 rows, cols;
+
+  if (cockpit_json_get_int (options, "window_rows", default_rows, &rows) &&
```
Comment (stefwalter) [inline src/bridge/cockpitpipechannel.c]: This should be documented in doc/protocol.md

In addition I would suggest: 

```
{
   "window": {
       "rows": 80,
       "cols": 25,
    }
}
```

File: src/bridge/cockpitpipechannel.c [hunk 4]
```diff
@@ -105,6 +107,27 @@ process_pipe_buffer (CockpitPipeChannel *self,
...
+      cockpit_json_get_int (options, "window_cols", default_cols, &cols) &&
+      rows >= 0 && rows < G_MAXUINT16 && cols >= 0 && cols < G_MAXUINT16)
+    {
```
Comment (stefwalter) [inline src/bridge/cockpitpipechannel.c]: When setting an unsupported value here, we should do a g_message() about it. In addition these limitations shuold be documented in doc/protocol.md when you document the window.rows and window.cols options.

File: src/bridge/cockpitpipechannel.c [hunk 5]
```diff
@@ -130,6 +153,27 @@ cockpit_pipe_channel_control (CockpitChannel *channel,
...
+              goto out;
+            }
+
+          if (rows > 0 && cols > 0)
+            {
+              struct winsize size = { rows, cols, 0, 0 };
+              gint fd;
+
+              g_object_get (self->pipe, "in-fd", &fd, NULL);
+              ioctl (fd, TIOCSWINSZ, &size);
```
Comment (stefwalter) [inline src/bridge/cockpitpipechannel.c]: We should print out with a g_warning() here (if we expect this never to fail on a valid system).

File: src/common/test-pipe.c [hunk 6]
```diff
@@ -833,7 +833,7 @@ test_pty_shell (void)
 
   const gchar *argv[] = { "/bin/bash", "-i", NULL };
 
-  pipe = cockpit_pipe_pty (argv, NULL, NULL);
+  pipe = cockpit_pipe_pty (argv, NULL, NULL, 24, 80);
```
Comment (stefwalter) [inline src/common/test-pipe.c]: Is there a way we can unit test:

 * Opening a channel with the right options.
 * Changing the options inside an already open channel.

I believe that spawning a shell from test-pipe and having the shell send out its ROWS and COLS environment variables after receiving some trigger input ... that would do the trick no?

**Comment by andreasn** [comment]
If I make the window smaller, and then bigger again I seem to lose some information.
screenshot from 2016-06-27 12-43-08

**Comment by andreasn** [comment]
I was pondering if it needs a switch to switch between this fullscreen mode and the 80x56 (or whatever the standard is), but I can't really find any good rationale for keeping the 80x56 mode. Does anyone else?

File: pkg/base1/cockpit.css (hunk 1 above)
Comment (larsu) [inline pkg/base1/cockpit.css]: It's relative to the page. The 44px is the header, which is also hard-coded to be 44px high.

This is the best way I've found to determine the height of the page. I'm very open to changing that if there's a better way.

File: pkg/base1/term.css (hunk 2 above)
Comment (larsu) [inline pkg/base1/term.css]: Those have always been overridden by the ones in cockpit.js.

File: src/bridge/cockpitpipechannel.c (hunk 3 above)
Comment (larsu) [inline src/bridge/cockpitpipechannel.c]: I agree. Will change that

**Comment by larsu** [comment]
> If I make the window smaller, and then bigger again I seem to lose some information.

Ya. This is a limitation of term.js. I'm not sure if it's worth to fix that, as it would be quite some work on the internals of that. The alternative is to not make it fullscreen, but just bigger. That could lead to another level of scrolling for some window sizes, though.

> I was pondering if it needs a switch to switch between this fullscreen mode and the 80x56 (or whatever the standard is), but I can't really find any good rationale for keeping the 80x56 mode. Does anyone else?

I agree that fullscreen is enough.

**Comment by larsu** [comment]
> If I make the window smaller, and then bigger again I seem to lose some information.

Ya. This is a limitation of term.js. I'm not sure if it's worth to fix that, as it would be quite some work on the internals of that. The alternative is to not make it fullscreen, but just bigger. That could lead to another level of scrolling for some window sizes, though.

> I was pondering if it needs a switch to switch between this fullscreen mode and the 80x56 (or whatever the standard is), but I can't really find any good rationale for keeping the 80x56 mode. Does anyone else?

I agree that fullscreen is enough.

**Comment by larsu** [comment]
Update: change the options to window.cols and .rows and document them; throw a warning when setting the ioctl fails; add a test.

**Comment by stefwalter** [comment]
This test failure seems related:

```
# testBasic (check_terminal.TestTerminal)
#
Wrote TestTerminal-testBasic-FAIL.png
Journal extracted to TestTerminal-testBasic-10.111.118.244-FAIL.log
not ok 72 testBasic (check_terminal.TestTerminal)
Traceback (most recent call last):
  File "./verify/check-terminal", line 64, in testBasic
    wait_line(2, "admin")
  File "./verify/check-terminal", line 46, in wait_line
    b.wait_text(line_sel(i), line_text(t))
  File "/build/cockpit/test/common/testlib.py", line 246, in wait_text
    return self.wait_js_func('ph_text_is', selector, text)
  File "/build/cockpit/test/common/testlib.py", line 210, in wait_js_func
    return self.phantom.wait("%s(%s)" % (func, ','.join(map(jsquote, args))))
  File "/build/cockpit/test/common/testlib.py", line 664, in <lambda>
    return lambda *args: self._invoke(name, *args)
  File "/build/cockpit/test/common/testlib.py", line 687, in _invoke
    raise Error(res['error'])
Error: timeout
```

**Comment by larsu** [comment]
> This test failure seems related:

Indeed, it assumed that the terminal was always 80 columns wide. Fixed.

File: lib/term.css [hunk 7]
```diff
@@ -1,17 +1,3 @@
-.terminal {
-  float: left;
-  border: #000000 solid 5px;
-  font-family: "DejaVu Sans Mono", "Liberation Mono", monospace;
-  font-size: 11px;
-  color: #f0f0f0;
-  background: #000000;
-}
-
-.terminal-cursor {
```
Comment (stefwalter) [inline lib/term.css]: Why was this removed?

File: src/bridge/cockpitpipechannel.c [hunk 8]
```diff
@@ -105,6 +108,38 @@ process_pipe_buffer (CockpitPipeChannel *self,
...
+  if (window == NULL)
+    {
+      *rowsp = default_rows;
+      *colsp = default_cols;
+      return TRUE;
+    }
+
+  if (cockpit_json_get_int (window, "rows", default_rows, &rows) &&
+      cockpit_json_get_int (window, "cols", default_cols, &cols) &&
+      rows >= 0 && rows <= G_MAXUINT16 && cols >= 0 && cols <= G_MAXUINT16)
```
Comment (stefwalter) [inline src/bridge/cockpitpipechannel.c]: I know it's a long shot, but the front end doesn't seem to limit itself to G_MAXUINT16? How about just clamping it to these maximums, rather than disconnecting the front end for passing invalid data?

File: lib/term.css (hunk 7 above)
Comment (larsu) [inline lib/term.css]: Because cockpit.css has the similar rules which already overrode these [1]. Would you prefer to remove the ones there and keep them here in term.css?

[1] https://github.com/cockpit-project/cockpit/blob/master/src/base1/cockpit.css#L315

File: src/bridge/cockpitpipechannel.c [hunk 9]
```diff
@@ -130,6 +165,28 @@ cockpit_pipe_channel_control (CockpitChannel *channel,
...
         }
 
+      /* ignore size options if this channel is not a pty or we are in prepare() */
+      if (self->pty && self->pipe)
+        {
+          gushort rows, cols;
+
+          if (!cockpit_pipe_channel_read_window_size_options (message, 0, 0, &rows, &cols))
+            {
+              g_warning ("invalid \"window.rows\" or \"window.cols\" option for stream channel");
```
Comment (stefwalter) [inline src/bridge/cockpitpipechannel.c]: Could we include the channel name for a bit easier debugging? That is ```g_warning("%s: invalid ...", self->name);``` and similar in the message below.

File: src/bridge/cockpitpipechannel.c [hunk 10]
```diff
@@ -130,6 +165,28 @@ cockpit_pipe_channel_control (CockpitChannel *channel,
...
+              goto out;
+            }
+
+          if (rows > 0 && cols > 0)
+            {
+              struct winsize size = { rows, cols, 0, 0 };
+              gint fd;
+
+              g_object_get (self->pipe, "in-fd", &fd, NULL);
+              if (ioctl (fd, TIOCSWINSZ, &size) < 0)
```
Comment (stefwalter) [inline src/bridge/cockpitpipechannel.c]: If fd is -1 then in-fd is already closed, lets not call ioctl in that case.

File: src/common/cockpittransport.c [hunk 11]
```diff
@@ -165,6 +165,27 @@ cockpit_transport_emit_recv (CockpitTransport *transport,
 }
 
 void
+cockpit_transport_emit_control (CockpitTransport *transport,
```
Comment (stefwalter) [inline src/common/cockpittransport.c]: Should cockpit_transport_default_recv() call this function? Otherwise we have code duplication.

**Comment by stefwalter** [comment]
screenshot from 2016-07-08 08-59-43

I get scrollbars on the terminal window. 

screenshot from 2016-07-08 09-02-42

The kubernetes log window is also displaying strangely.

File: src/bridge/cockpitpipechannel.c (hunk 8 above)
Comment (larsu) [inline src/bridge/cockpitpipechannel.c]: I agree, thanks.

File: src/bridge/cockpitpipechannel.c (hunk 9 above)
Comment (larsu) [inline src/bridge/cockpitpipechannel.c]: Yep.

**Comment by stefwalter** [comment]
screenshot from 2016-07-08 09-13-06

Ditto for containers displayed without logs.

File: src/base1/cockpit.css [hunk 12]
```diff
@@ -313,15 +313,27 @@ a {
...
 /* Our terminal or logs */
+.console-ct-container {
+    border: none;
+}
+
 .console-ct {
     font-family: Menlo, Monaco, Consolas, monospace;
-    margin-top: 0;
-    margin-bottom: 0;
+    background-color: black;
```
Comment (stefwalter) [inline src/base1/cockpit.css]: console-ct is used by both terminals and log output displays. See the screenshots of the kubernetes and docker logs attached to this bug. I don't think this should have a black background.

**Comment by larsu** [comment]
Thanks for the comments. Addressed all of them except:

> I get scrollbars on the terminal window.

How do you get those? I'm afraid I can't reproduce that and I don't see them in your screenshot either.

**Comment by stefwalter** [comment]
```
src/common/cockpittransport.c: In function 'cockpit_transport_default_recv':
src/common/cockpittransport.c:61:12: error: unused variable 'ret' [-Werror=unused-variable]
   gboolean ret = FALSE;
```

**Comment by larskarlitski** [comment]
Rebased and updated. Would appreciate another look :)

**Comment by martinpitt** [comment]
The tests all fail with
```
Module not found: Error: Cannot resolve 'file' or 'directory' ../pkg/systemd/terminal.css in /build/cockpit/tmp-dist
resolve file
  /build/cockpit/pkg/systemd/terminal.css doesn't exist
  /build/cockpit/pkg/systemd/terminal.css.webpack.js doesn't exist
  /build/cockpit/pkg/systemd/terminal.css.web.js doesn't exist
  /build/cockpit/pkg/systemd/terminal.css.js doesn't exist
  /build/cockpit/pkg/systemd/terminal.css.json doesn't exist
```

**Comment by larskarlitski** [comment]
> The tests all fail with

Of course they do. Forgot a file. Sorry and thanks!

**Comment by larskarlitski** [comment]
The `check-terminal` test failed on Ubuntu, because it has a different sudo help text which it also shows after clicking "Reset".

Updated to make the text more robust (we were hard coding 4 lines of sudo message before).

File: src/bridge/cockpitpipechannel.c [hunk 13]
```diff
@@ -401,7 +461,7 @@ cockpit_pipe_channel_prepare (CockpitChannel *channel)
                                 "invalid \"directory\" option for stream channel");
           goto out;
         }
-      if (!cockpit_json_get_bool (options, "pty", FALSE, &pty))
+      if (!cockpit_json_get_bool (options, "pty", FALSE, &self->pty))
```
Comment (martinpitt) [inline src/bridge/cockpitpipechannel.c]: Smells like the `gboolean pty` function-local variable should be dropped now?

File: src/common/test-pipe.c (hunk 6 above)
Comment (martinpitt) [inline src/common/test-pipe.c]: The following commit now has exactly such a test case.

File: pkg/lib/cockpit-components-terminal.jsx [hunk 14]
```diff
@@ -94,8 +100,10 @@
         },
 
         render: function () {
+            var cls = 'console-ct';
+
             // ensure react never reuses this div by keying it with the terminal widget
-            return <div ref="terminal" className="console-ct" key={this.state.terminal} />;
+            return <div ref="terminal" className={cls} key={this.state.terminal} />;
```
Comment (martinpitt) [inline pkg/lib/cockpit-components-terminal.jsx]: What's the purpose of the new `cls`? As far as I can see it's being used just once, right here. Debugging leftover?

File: pkg/lib/cockpit-components-terminal.jsx [hunk 15]
```diff
@@ -132,6 +140,25 @@
...
+            ch.textContent = 'M';
+            terminal.appendChild(ch);
+            var height = ch.offsetHeight; // offsetHeight is only correct for block elements
+            ch.style.display = 'inline';
+            var width = ch.offsetWidth;
+            terminal.removeChild(ch);
+
+            this.setState({
+                rows: Math.floor((node.parentElement.clientHeight - padding) / height),
+                cols: Math.floor((node.parentElement.clientWidth - padding)/ width)
```
Comment (martinpitt) [inline pkg/lib/cockpit-components-terminal.jsx]: nitpick: missing space before /

File: pkg/lib/console.css [hunk 16]
```diff
@@ -36,15 +36,10 @@
     background-color: black;
     border: 1px solid black;
     padding: 10px;
-    display: inline-block; /* size DIV to contents */
+    /* display: inline-block; */ /* size DIV to contents */
```
Comment (martinpitt) [inline pkg/lib/console.css]: debugging leftover?

**Comment by andreasn** [comment]
Looks good to me!

**Comment by larskarlitski** [comment]
@pitti, thanks! Fixed both.

**Comment by pitti** [comment]
@larskarlitski @martinpitt 

No problem, always happy to help! :)

File: src/bridge/cockpitpipechannel.c (hunk 13 above)
Comment (larskarlitski) [inline src/bridge/cockpitpipechannel.c]: I actually did drop it, but in a later commit. Moved it into the one that it belongs to. Thanks!

File: pkg/lib/cockpit-components-terminal.jsx (hunk 14 above)
Comment (larskarlitski) [inline pkg/lib/cockpit-components-terminal.jsx]: Yeah, this was a leftover from a previous version.

**Comment by larskarlitski** [comment]
Rebased again taking into account @martinpitt's other suggestions.

**Comment by martinpitt** [comment]
All green, let's land this at last ☺

**Comment by PeterFalken** [comment]
Thanks Martin - looking forward to using it on our servers.

**Comment by martinpitt** [comment]
Current screenshot, for release notes:

terminal-full-window
